File: 14-LongestCommonPrefix/14-LongestCommonPrefix.py
# Last updated: 8/16/2026, 9:53:41 PM
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        # if len(strs) == 0: return ''
        # if len(strs) == 1: return strs[0]
        # res = []
        # size = len(strs)

        # for i in range(len(strs[0])):
        #     for j in range(size - 1):
        #         flag = False
        #         if len(strs[j]) > i and len(strs[j + 1]) > i and strs[j][i] ==  strs[j+1][i]:
        #             flag = True
        #         else:
        #             break
        #     if flag:
        #         res.append(strs[j][i])
        #     else:
        #         break
        
        # return ''.join(res)

        # no need for flag, if not match, directly return current visited strs[0]
        # for strs[0], check each char with following items
        if len(strs) == 0: return ''

        for i in range(len(strs[0])):
            c = strs[0][i]
            for item in strs[1:]:
                if i >= len(item) or c != item[i]:
                    # return so far visited chars in strs[0]
                    return strs[0][:i]
        # finish iterating strs[0], every char matched
        return strs[0]

                

File: 14-LongestCommonPrefix/14-LongestCommonPrefix.java
// Last updated: 8/16/2026, 9:53:44 PM
class Solution {
    public String longestCommonPrefix(String[] strs) {
        // compare first string with remaining strings
        if(strs.length == 0) return "";
        if(strs.length == 1) return strs[0];
        String s1 = strs[0];
        int len1 = s1.length();
        StringBuilder res = new StringBuilder(len1); // LCA: max is the s1
        
        for(int j = 0; j < len1; j ++) {
            // compare string with s1, by char at index j
            char c = s1.charAt(j);
            for(int i = 1; i < strs.length; i ++) {
                // compare s1 with remaining strings

                if(strs[i].length() <= j || c != strs[i].charAt(j)) {
                    return res.toString();
                } 
            }
            res.append(s1.charAt(j));
        }
        return res.toString();
    }
}